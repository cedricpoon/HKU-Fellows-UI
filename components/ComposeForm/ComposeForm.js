import React, { PureComponent } from 'react';
import { Linking } from 'react-native';
import { Content, Form, Item, Input, Textarea, Button, Text, View } from 'native-base';
import PropTypes from 'prop-types';

import { markdownTutorialLink } from 'hkufui/config';
import { localize } from 'hkufui/locale';
import styles from './Styles';

const locale = localize({ country: 'hk', language: 'en' });

class ComposeForm extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { isNative: true };
    this._toggleUploadMode = this._toggleUploadMode.bind(this);
  }

  _toggleUploadMode() {
    const { onToggleMode } = this.props;
    this.setState(prevState => {
      onToggleMode(!prevState.isNative);
      return ({ isNative: !prevState.isNative });
    });
  }

  render() {
    const { onTextUpdates, screenHeight, onToggleMode, replyParams } = this.props;
    const { isNative } = this.state;

    return (
      <Content padder>
        <Form>
          <Item regular style={replyParams && styles.replybox}>
            <Input
              placeholder={locale['new.title']}
              style={[styles.textbox, replyParams && styles.bold]}
              onChangeText={onTextUpdates.title}
              disabled={replyParams != null}
              value={replyParams ? `${locale['reply.titlePrefix']}${replyParams.title}` : null}
            />
          </Item>
          <Item regular style={[styles.item, !isNative && styles.hidden, replyParams && styles.replybox]}>
            <Input
              placeholder={!replyParams ? locale['new.subtitle'] : locale['post.noSubtitle']}
              style={styles.textbox}
              onChangeText={onTextUpdates.subtitle}
              disabled={replyParams != null}
              value={replyParams ? replyParams.subtitle : null}
            />
          </Item>
          {!replyParams && (
            <Item regular style={[styles.item, !isNative && styles.hidden]}>
              <Input placeholder={locale['new.hashtags']} style={styles.textbox} onChangeText={onTextUpdates.hashtags} />
            </Item>
          )}
          <Textarea
            bordered
            placeholder={locale['new.content']}
            style={[{ height: screenHeight / 3 }, styles.item]}
            onChangeText={onTextUpdates.content}
          />
          {onToggleMode && (
            <Button full dark={!isNative} light={isNative} style={styles.item} onPress={this._toggleUploadMode}>
              <Text style={styles.toggler}>{isNative ? 'As Native Post' : 'As Moodle Forum Post'}</Text>
            </Button>
          )}
        </Form>
        <View style={styles.remarkGroup}>
          <Text style={styles.remark} note>
            {locale['new.markdownRemark']}
          </Text>
          <Text
            style={[styles.remark, styles.hyperlink]}
            onPress={() => Linking.openURL(markdownTutorialLink)}
          >
            {locale['new.markdown']}
          </Text>
        </View>
      </Content>
    );
  }
}

ComposeForm.defaultProps = {
  onTextUpdates: {}
}

ComposeForm.propTypes = {
  onTextUpdates: PropTypes.shape({
    title: PropTypes.func,
    subtitle: PropTypes.func,
    hashtags: PropTypes.func,
    content: PropTypes.func
  }),
  onToggleMode: PropTypes.func,
  screenHeight: PropTypes.number.isRequired,
  replyParams: PropTypes.shape({
    title: PropTypes.string.isRequired,
    subtitle: PropTypes.string,
    native: PropTypes.bool
  })
}

export default ComposeForm;
