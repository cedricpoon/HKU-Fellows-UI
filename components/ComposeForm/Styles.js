import { StyleSheet } from 'react-native';
import { gridBase, fontSize } from 'hkufui/theme/grid';
import { logo } from 'hkufui/theme/palette';

export default StyleSheet.create({
  textbox: {
    fontSize: fontSize.titleSize,
    height: gridBase * 6,
    paddingLeft: gridBase * 2,
    paddingRight: gridBase
  },
  replybox: {
    borderTopWidth: 0,
    borderBottomWidth: 0,
    borderRightWidth: 0,
    borderLeftWidth: 2
  },
  bold: {
    fontWeight: "600"
  },
  item: {
    marginTop: gridBase * 2
  },
  remarkGroup: {
    marginTop: gridBase * 2,
    flexDirection: 'row'
  },
  remark: {
    fontSize: fontSize.semiTitleSize,
    fontStyle: 'italic',
    marginRight: gridBase
  },
  hyperlink: {
    color: logo.blue,
    fontStyle: 'normal',
    textDecorationLine: 'underline'
  },
  toggler: {
    fontSize: fontSize.titleSize
  },
  hidden: {
    display: 'none'
  }
});
