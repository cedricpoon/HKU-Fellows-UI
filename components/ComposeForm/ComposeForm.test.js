import React from 'react';
import { shallow } from 'enzyme';
import ComposeForm from './ComposeForm';

describe('Testing ComposeForm Component', () => {
  const defaultProps = {
    screenHeight: 300
  };

  it('renders as expected', () => {
    const wrapper = shallow(
      <ComposeForm {...defaultProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with onTextUpdates', () => {
    const wrapper = shallow(
      <ComposeForm {...defaultProps} onTextUpdates={{ title: () => {} }} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('can toggle mode', () => {
    const wrapper = shallow(
      <ComposeForm {...defaultProps} onToggleMode={() => {}} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
