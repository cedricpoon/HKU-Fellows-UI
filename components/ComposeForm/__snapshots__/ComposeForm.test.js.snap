// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ComposeForm Component can toggle mode 1`] = `
<Styled(Content)
  padder={true}
>
  <Styled(Form)>
    <Styled(Item)
      regular={true}
    >
      <Styled(Input)
        disabled={false}
        placeholder="Title (Compulsory)"
        style={
          Array [
            Object {
              "fontSize": 15,
              "height": 30,
              "paddingLeft": 10,
              "paddingRight": 5,
            },
            undefined,
          ]
        }
        value={null}
      />
    </Styled(Item)>
    <Styled(Item)
      regular={true}
      style={
        Array [
          Object {
            "marginTop": 10,
          },
          false,
          undefined,
        ]
      }
    >
      <Styled(Input)
        disabled={false}
        placeholder="Subtitle"
        style={
          Object {
            "fontSize": 15,
            "height": 30,
            "paddingLeft": 10,
            "paddingRight": 5,
          }
        }
        value={null}
      />
    </Styled(Item)>
    <Styled(Item)
      regular={true}
      style={
        Array [
          Object {
            "marginTop": 10,
          },
          false,
        ]
      }
    >
      <Styled(Input)
        placeholder="Hashtags (Max. 2, E.g. #AY18#Tutor1)"
        style={
          Object {
            "fontSize": 15,
            "height": 30,
            "paddingLeft": 10,
            "paddingRight": 5,
          }
        }
      />
    </Styled(Item)>
    <Styled(Textarea)
      bordered={true}
      placeholder="Content (Compulsory)"
      style={
        Array [
          Object {
            "height": 100,
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
    />
    <Styled(Button)
      dark={false}
      full={true}
      light={true}
      onPress={[Function]}
      style={
        Object {
          "marginTop": 10,
        }
      }
    >
      <Styled(Text)
        style={
          Object {
            "fontSize": 15,
          }
        }
      >
        As Native Post
      </Styled(Text)>
    </Styled(Button)>
  </Styled(Form)>
  <Styled(ViewNB)
    style={
      Object {
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <Styled(Text)
      note={true}
      style={
        Object {
          "fontSize": 12.5,
          "fontStyle": "italic",
          "marginRight": 5,
        }
      }
    >
      ** Content can be styled in
    </Styled(Text)>
    <Styled(Text)
      onPress={[Function]}
      style={
        Array [
          Object {
            "fontSize": 12.5,
            "fontStyle": "italic",
            "marginRight": 5,
          },
          Object {
            "color": "#3B96DC",
            "fontStyle": "normal",
            "textDecorationLine": "underline",
          },
        ]
      }
    >
      Markdown
    </Styled(Text)>
  </Styled(ViewNB)>
</Styled(Content)>
`;

exports[`Testing ComposeForm Component is a reply 1`] = `
<Styled(Content)
  padder={true}
>
  <Styled(Form)>
    <Styled(Item)
      regular={true}
      style={
        Object {
          "borderBottomWidth": 0,
          "borderLeftWidth": 2,
          "borderRightWidth": 0,
          "borderTopWidth": 0,
        }
      }
    >
      <Styled(Input)
        disabled={true}
        placeholder="Title (Compulsory)"
        style={
          Array [
            Object {
              "fontSize": 15,
              "height": 30,
              "paddingLeft": 10,
              "paddingRight": 5,
            },
            Object {
              "fontWeight": "600",
            },
          ]
        }
        value="Re: title"
      />
    </Styled(Item)>
    <Styled(Item)
      regular={true}
      style={
        Array [
          Object {
            "marginTop": 10,
          },
          false,
          Object {
            "borderBottomWidth": 0,
            "borderLeftWidth": 2,
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
        ]
      }
    >
      <Styled(Input)
        disabled={true}
        placeholder="- No Subtitle -"
        style={
          Object {
            "fontSize": 15,
            "height": 30,
            "paddingLeft": 10,
            "paddingRight": 5,
          }
        }
      />
    </Styled(Item)>
    <Styled(Textarea)
      bordered={true}
      placeholder="Content (Compulsory)"
      style={
        Array [
          Object {
            "height": 100,
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
    />
  </Styled(Form)>
  <Styled(ViewNB)
    style={
      Object {
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <Styled(Text)
      note={true}
      style={
        Object {
          "fontSize": 12.5,
          "fontStyle": "italic",
          "marginRight": 5,
        }
      }
    >
      ** Content can be styled in
    </Styled(Text)>
    <Styled(Text)
      onPress={[Function]}
      style={
        Array [
          Object {
            "fontSize": 12.5,
            "fontStyle": "italic",
            "marginRight": 5,
          },
          Object {
            "color": "#3B96DC",
            "fontStyle": "normal",
            "textDecorationLine": "underline",
          },
        ]
      }
    >
      Markdown
    </Styled(Text)>
  </Styled(ViewNB)>
</Styled(Content)>
`;

exports[`Testing ComposeForm Component renders as expected 1`] = `
<Styled(Content)
  padder={true}
>
  <Styled(Form)>
    <Styled(Item)
      regular={true}
    >
      <Styled(Input)
        disabled={false}
        placeholder="Title (Compulsory)"
        style={
          Array [
            Object {
              "fontSize": 15,
              "height": 30,
              "paddingLeft": 10,
              "paddingRight": 5,
            },
            undefined,
          ]
        }
        value={null}
      />
    </Styled(Item)>
    <Styled(Item)
      regular={true}
      style={
        Array [
          Object {
            "marginTop": 10,
          },
          false,
          undefined,
        ]
      }
    >
      <Styled(Input)
        disabled={false}
        placeholder="Subtitle"
        style={
          Object {
            "fontSize": 15,
            "height": 30,
            "paddingLeft": 10,
            "paddingRight": 5,
          }
        }
        value={null}
      />
    </Styled(Item)>
    <Styled(Item)
      regular={true}
      style={
        Array [
          Object {
            "marginTop": 10,
          },
          false,
        ]
      }
    >
      <Styled(Input)
        placeholder="Hashtags (Max. 2, E.g. #AY18#Tutor1)"
        style={
          Object {
            "fontSize": 15,
            "height": 30,
            "paddingLeft": 10,
            "paddingRight": 5,
          }
        }
      />
    </Styled(Item)>
    <Styled(Textarea)
      bordered={true}
      placeholder="Content (Compulsory)"
      style={
        Array [
          Object {
            "height": 100,
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
    />
  </Styled(Form)>
  <Styled(ViewNB)
    style={
      Object {
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <Styled(Text)
      note={true}
      style={
        Object {
          "fontSize": 12.5,
          "fontStyle": "italic",
          "marginRight": 5,
        }
      }
    >
      ** Content can be styled in
    </Styled(Text)>
    <Styled(Text)
      onPress={[Function]}
      style={
        Array [
          Object {
            "fontSize": 12.5,
            "fontStyle": "italic",
            "marginRight": 5,
          },
          Object {
            "color": "#3B96DC",
            "fontStyle": "normal",
            "textDecorationLine": "underline",
          },
        ]
      }
    >
      Markdown
    </Styled(Text)>
  </Styled(ViewNB)>
</Styled(Content)>
`;

exports[`Testing ComposeForm Component renders with onTextUpdates 1`] = `
<Styled(Content)
  padder={true}
>
  <Styled(Form)>
    <Styled(Item)
      regular={true}
    >
      <Styled(Input)
        disabled={false}
        onChangeText={[Function]}
        placeholder="Title (Compulsory)"
        style={
          Array [
            Object {
              "fontSize": 15,
              "height": 30,
              "paddingLeft": 10,
              "paddingRight": 5,
            },
            undefined,
          ]
        }
        value={null}
      />
    </Styled(Item)>
    <Styled(Item)
      regular={true}
      style={
        Array [
          Object {
            "marginTop": 10,
          },
          false,
          undefined,
        ]
      }
    >
      <Styled(Input)
        disabled={false}
        placeholder="Subtitle"
        style={
          Object {
            "fontSize": 15,
            "height": 30,
            "paddingLeft": 10,
            "paddingRight": 5,
          }
        }
        value={null}
      />
    </Styled(Item)>
    <Styled(Item)
      regular={true}
      style={
        Array [
          Object {
            "marginTop": 10,
          },
          false,
        ]
      }
    >
      <Styled(Input)
        placeholder="Hashtags (Max. 2, E.g. #AY18#Tutor1)"
        style={
          Object {
            "fontSize": 15,
            "height": 30,
            "paddingLeft": 10,
            "paddingRight": 5,
          }
        }
      />
    </Styled(Item)>
    <Styled(Textarea)
      bordered={true}
      placeholder="Content (Compulsory)"
      style={
        Array [
          Object {
            "height": 100,
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
    />
  </Styled(Form)>
  <Styled(ViewNB)
    style={
      Object {
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <Styled(Text)
      note={true}
      style={
        Object {
          "fontSize": 12.5,
          "fontStyle": "italic",
          "marginRight": 5,
        }
      }
    >
      ** Content can be styled in
    </Styled(Text)>
    <Styled(Text)
      onPress={[Function]}
      style={
        Array [
          Object {
            "fontSize": 12.5,
            "fontStyle": "italic",
            "marginRight": 5,
          },
          Object {
            "color": "#3B96DC",
            "fontStyle": "normal",
            "textDecorationLine": "underline",
          },
        ]
      }
    >
      Markdown
    </Styled(Text)>
  </Styled(ViewNB)>
</Styled(Content)>
`;
